### PR TITLE
Fixes #3232 by handling terminating status updates for resident tasks:

### LINKED ISSUE #3232: Resident Tasks: Treat status updates correctly
Task status updates should be treated correctly. Probably similar to the following:

On termination, remove the "launch" information from the task. If the task has a reservation, keep it, otherwise delete it.

### PR DESCRIPTION
**Pull request by aquamatthias**
- introduce an Unlaunch operation
- Handling Unlaunch by removing the task launch information

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala [hunk 1]
```diff
@@ -39,40 +40,55 @@ private[tracker] object TaskOpProcessorImpl {
...
+          log.warn("Got update for task which wasn't launched yet: {}", statusUpdate)
+          Action.Noop
+        }
+    }
+    private[this] def resolveForResidentFailed(task: Task, statusUpdate: TaskStatus): Option[Action] = {
+      for {
+        reservation <- task.reservationWithVolumes
+        launched <- task.launched
+        if Terminated.isTerminated(statusUpdate.getState)
+      } yield Unlaunch(task)
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: Please update the scaladoc on line 37 – it should mention the possibility of this action as well.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala [hunk 2]
```diff
@@ -39,40 +40,55 @@ private[tracker] object TaskOpProcessorImpl {
...
         case Some(existingTask) =>
-          resolveForExistingTask(existingTask, status)
+          actionForTaskAndStatus(existingTask, status)
         case None =>
           Action.Fail(new IllegalStateException(s"$taskId of app [${taskId.appId}] does not exist"))
       }
     }
 
+    private[this] def actionForTaskAndStatus(task: Task, statusUpdate: TaskStatus): Action = {
+      //Step 1: if a resident task is terminated, it has to be handled different
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: nit: _differently_, it's an adverb :P

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala [hunk 3]
```diff
@@ -39,40 +40,55 @@ private[tracker] object TaskOpProcessorImpl {
...
-                taskState.copy(launched = Some(
-                  currentLaunched.copy(
-                    status = currentLaunched.status.copy(
-                      startedAt = Some(now),
-                      mesosStatus = Some(statusUpdate)
-                    )
-                  )
-                ))
+    private[this] def resolveForLaunchedTask(task: Task, statusUpdate: TaskStatus): Option[Action] = {
+      def stagedNowRunning(currentLaunched: Task.Launched): Action = {
```
Comment (meichstedt) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: `stagedNowRunning` is defined in local scope of `resolveForLaunchedTask`, `updateTaskOnStateChange` is defined outside of this scope. Both are only used from within `resolveForLaunchedTask`, so I'd suggest to define both in the same scope (I don't really mind which one).

**Comment by meichstedt** [comment]
Nice! Just minor comments – would be great if you can address these.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 2 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: thanks.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 1 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: Done.

File: src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala (hunk 3 above)
Comment (aquamatthias) [inline src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala]: Defined the function outside the body - as the other one.

**Comment by aquamatthias** [comment]
@meichstedt Thanks for the hints! Please review.

**Comment by meichstedt** [comment]
Awesome – please squash!

**Comment by aquamatthias** [comment]
Squashed.

**Comment by meichstedt** [comment]
Thanks!
```
   __|=|_|=|__
   \ o.o.o.oY/
    \_______/
  ~~~~~~~~~~~~~~
```

**Pull request merged**